### PR TITLE
Add bundle cache

### PR DESCRIPTION
**Pull request by wu-hui**
Corresponding Android PRs (Not exact match):
https://github.com/firebase/firebase-android-sdk/pull/2301
https://github.com/firebase/firebase-android-sdk/pull/2332

Previous iOS PRs:
https://github.com/firebase/firebase-ios-sdk/pull/7283
https://github.com/firebase/firebase-ios-sdk/pull/7303

### CODE REVIEW

**Review by var-const — CHANGES_REQUESTED** [review]

File: Firestore/core/test/unit/local/bundle_cache_test.cc [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+#include "Firestore/core/src/core/target.h"
+#include "Firestore/core/src/local/bundle_cache.h"
+#include "Firestore/core/src/local/persistence.h"
+#include "Firestore/core/test/unit/testutil/testutil.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/unit/local/bundle_cache_test.cc:36]: Done.

File: Firestore/core/test/unit/local/bundle_cache_test.cc [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+using bundle::BundledQuery;
+using bundle::BundleMetadata;
+using bundle::NamedQuery;
+using core::Query;
+using core::Target;
+using model::SnapshotVersion;
+using testutil::Filter;
+
+BundleCacheTest::BundleCacheTest(std::unique_ptr<Persistence> persistence)
+    : persistence_(std::move(persistence)),
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/unit/local/bundle_cache_test.cc]: Done.

File: Firestore/core/test/unit/local/bundle_cache_test.cc [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+      cache_(persistence_->bundle_cache()) {
+}
+
+BundleCacheTest::BundleCacheTest() : BundleCacheTest(GetParam()()) {
+}
+
+BundleCacheTest::~BundleCacheTest() = default;
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/unit/local/bundle_cache_test.cc]: Done.

File: Firestore/core/test/unit/local/bundle_cache_test.cc [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+TEST_P(BundleCacheTest, ReturnsNullOptWhenBundleIdNotFound) {
+  persistence_->Run("test_returns_nullopt_when_bundle_id_not_found", [&] {
+    ASSERT_EQ(absl::nullopt, cache_->GetBundleMetadata("bundle-1"));
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/unit/local/bundle_cache_test.cc]: Done.

[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/unit/local/bundle_cache_test.cc]: Done.

File: Firestore/core/test/unit/local/bundle_cache_test.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2019 Google
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/unit/local/bundle_cache_test.h]: Done.

File: Firestore/core/test/unit/local/bundle_cache_test.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+class BundleCache;
+
+using FactoryFunc = std::unique_ptr<Persistence> (*)();
+
+/**
+ * These are tests for any implementation of the BundleCache interface.
+ *
+ * To test a specific implementation of TargetCache:
+ *
+ * + Write a persistence factory function
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/unit/local/bundle_cache_test.h]: Done.

File: Firestore/core/test/unit/local/local_serializer_test.cc [hunk 7]
```diff
@@ -52,7 +57,10 @@ namespace firestore {
 namespace local {
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/unit/local/local_serializer_test.cc:59]: Done.

File: Firestore/core/test/unit/local/bundle_cache_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/test/unit/local/bundle_cache_test.cc:36]: Sorry, to be clear -- pretty much everything in the test file should be in an unnamed namespace. That includes any helper classes, using declarations, free functions, etc. In most cases, the unnamed namespace should start immediately after the most-nested named namespace:
```cpp
namespace firebase {
namespace firestore {
namespace local {
namespace {
```

File: Firestore/core/test/unit/local/bundle_cache_test.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/test/unit/local/bundle_cache_test.cc]: Nit: please include `util/hard_assert.h`.

File: Firestore/core/src/local/leveldb_bundle_cache.cc [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+using bundle::BundledQuery;
+using bundle::BundleMetadata;
+using bundle::NamedQuery;
+using nanopb::ByteString;
+using nanopb::Message;
+using nanopb::StringReader;
+
+LevelDbBundleCache::LevelDbBundleCache(LevelDbPersistence* db,
+                                       LocalSerializer* serializer)
+    : db_(db), serializer_(serializer) {
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/local/leveldb_bundle_cache.cc]: Please include the appropriate header for `NOT_NULL` (`util/hard_assert.h`), it's best to not rely on this header being transitively included from somewhere.

**Review by var-const — APPROVED** [review]
LGTM (with just a couple of remaining nits).

File: Firestore/core/src/remote/serializer.cc [hunk 9]
```diff
@@ -1041,10 +1033,23 @@ Target Serializer::DecodeQueryTarget(
     end_at = DecodeBound(reader, query.end_at);
   }
 
-  return Query(std::move(path), std::move(collection_group),
-               std::move(filter_by), std::move(order_by), limit,
-               LimitType::First, std::move(start_at), std::move(end_at))
-      .ToTarget();
+  return Target(std::move(path), std::move(collection_group),
```
Comment (schmidt-sebastian) [inline Firestore/core/src/remote/serializer.cc:1036]: You should also reject Targets that have `offset` or `select` set.

**Pull request merged**